### PR TITLE
Add proxy support

### PR DESCRIPTION
**Pull request by abhipsaMisra**
Issue: #477 - Proxy support for Azure IoT Hub and Provisioning SDKs

With the release of .NET Core 2.1, we are implementing proxy support for the C# sdk. 
Currently, the sdk picks up the proxy settings as per the OS default proxy settings. With this change, we will be exposing an `IWebProxy` member to all our IoT SDK clients, allowing for setting the proxy explicitly in the application.

The following 3 scenarios will be supported:

1. Proxy not set in the code (sdk will pick the OS default proxy settings)
2. Proxy = null (sdk will not use proxy)
3. Proxy = set in the code (sdk will use custom proxy specified).

#### Device Client, Module Client
```C# 
using System.Net;

// s_proxyServerAddress = ...
// s_deviceConnectionString = …
// s_moduleConnectionString = ...
Http1TransportSettings transportSettings = new Http1TransportSettings();
transportSettings.Proxy = new WebProxy(s_proxyServerAddress);
ITransportSettings[] transportSettings = new ITransportSettings[] { transportSettings };

DeviceClient deviceClient = DeviceClient.CreateFromConnectionString(s_deviceConnectionString, transportSettings);
ModuleClient moduleClient = ModuleClient.CreateFromConnectionString(s_moduleConnectionString, transportSettings);
```
#### Job Client, Registry Manager
```C#
using System.Net;
// s_proxyServerAddress= ...
// s_connectionString = ...
HttpTransportSettings customTransportSettings = new HttpTransportSettings();
customTransportSettings.Proxy = new WebProxy(s_proxyServerAddress);

JobClient jobClient = JobClient.CreateFromConnectionString(s_connectionString, customTransportSettings);
RegistryManager registryManager = RegistryManager.CreateFromConnectionString(s_connectionString, customTransportSettings);
```
#### Service Client
```C#
using System.Net;
// s_proxyServerAddressAmqp = ...
// s_proxyServerAddressHttp = …
// s_connectionString = ...
ServiceClientTransportSettings transportSettings = new ServiceClientTransportSettings();
transportSettings.AmqpProxy = new WebProxy(s_proxyServerAddressAmqp);
transportSettings.HttpProxy = new WebProxy(s_proxyServerAddressHttp);

ServiceClient serviceClient = ServiceClient.CreateFromConnectionString(s_connectionString, transportSettings);
```
#### Provisioning Client
```C#
using System.Net;
// s_proxyServerAddress = ...
// GlobalDeviceEndpoint = ...
// s_idScope = ...
// security = ...
ProvisioningTransportHandlerHttp transport = new ProvisioningTransportHandlerHttp();
transport.Proxy = new WebProxy(s_proxyServerAddress);

ProvisioningDeviceClient provClient =ProvisioningDeviceClient.Create(GlobalDeviceEndpoint, s_idScope, security, transport);
```

Currently tested on Win 10 machine running .NET Core 2.1 (TFM .NET Core 2.1):
#### IoTHub Device Client:

  | Proxy = Not Set   (OS Default) | Proxy = null   (No proxy) | Proxy = Set   (Custom)
-- | -- | -- | --
HTTP Proxy | Yes | Yes | Yes
AMQP WS Proxy | Yes | Yes | Yes
MQTT WS Proxy | Yes | Yes | Yes
File Upload HTTP Proxy | Yes | No* | No*
* Azure storage currently does not have Proxy support for .NET Standard 2.0

#### IoTHub Service Client

  | Proxy = Not Set   (OS Default) | Proxy = null   (No proxy) | Proxy = Set   (Custom)
-- | -- | -- | --
IoTHub Service Client Proxy | Yes  | Yes  | Yes 
Registry Manager   HTTP Proxy | Yes | Yes | Yes
Jobs Client HTTP   Proxy | Yes | Yes | Yes
ADM Service HTTP   Proxy | Yes | Yes | Yes

#### Provisioning Device Client

  | Proxy = Not Set   (OS Default) | Proxy = null   (No proxy) | Proxy = Set   (Custom)
-- | -- | -- | --
HTTP Proxy | Yes | Yes | Yes
AMQP WS Proxy | Yes | Yes | Yes
MQTT WS Proxy | Yes | Yes | Yes

#### Provisioning Service Client

  | Proxy = Not Set   (OS Default) | Proxy = null   (No proxy) | Proxy = Set   (Custom)
-- | -- | -- | --
Provisioning Service HTTP Proxy | Yes | Yes | Yes

#### Module Client

  | Proxy = Not Set   (OS Default) | Proxy = null   (No proxy) | Proxy = Set   (Custom)
-- | -- | -- | --
AMQP WS Proxy | Yes | Yes | Yes
MQTT WS Proxy | Yes | Yes | Yes

## Checklist
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [x] This pull-request is submitted against the `master` branch.

## Description of the changes

- Added proxy support for device client

## Reference/Link to the issue solved with this PR (if any)

- Issue #477

### CODE REVIEW

**Comment by abhipsaMisra** [comment]
Currently tested on Win 10 machine running .NET Core 2.1 (TFM .NET Core 2.1):


  | Proxy = Not Set   (OS Default) | Proxy = null   (No proxy) | Proxy = Set   (Custom)
-- | -- | -- | --
HTTP Proxy | yes | yes | yes
AMQP WS Proxy | yes | yes | yes
MQTT WS Proxy | yes | yes | yes
File Upload HTTP   Proxy | yes | yes | yes

**Comment by kiranpradeep** [comment]
@abhipsaMisra Can't the proxy be set programatically as in C SDK[[1](https://github.com/Azure/azure-iot-sdk-c/blob/master/doc/Iothub_sdk_options.md#setting-an-option)]? If we are limited to accepting proxy from environment, then we will face issue[[2](https://github.com/Azure/azure-iot-sdk-c/issues/79)]  which we had faced in C sdk earlier.

**Comment by abhipsaMisra** [comment]
@kiranpradeep Thank you for pointing this out. I have updated the description, and have also added the sample usage of our implementation.

File: provisioning/service/src/Contract/Http1TransportSettings.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
```
Comment (CIPop) [inline provisioning/service/src/Contract/Http1TransportSettings.cs]: Missing copyright.

File: provisioning/service/src/ProvisioningServiceClient.cs [hunk 2]
```diff
@@ -857,6 +965,38 @@ public Query CreateEnrollmentGroupQuery(QuerySpecification querySpecification, i
...
+        ///     number of items per iteration can be specified by the pageSize. It is optional, you can provide <b>0</b> for
+        ///     default pageSize or use the API <see cref="CreateEnrollmentGroupQuery(QuerySpecification)"/>.
+        /// </remarks>
+        /// <param name="querySpecification">the <see cref="QuerySpecification"/> with the SQL query. It cannot be <code>null</code>.</param>
+        /// <param name="pageSize">the <code>int</code> with the maximum number of items per iteration. It can be 0 for default, but not negative.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <param name="httpTransportSettings"> Specifies the HTTP transport settings</param>
+        /// <returns>The <see cref="Query"/> iterator.</returns>
+        /// <exception cref="ArgumentException">if the provided parameters are not correct.</exception>
+        public Query CreateEnrollmentGroupQuery(QuerySpecification querySpecification, Http1TransportSettings httpTransportSettings, int pageSize, CancellationToken cancellationToken)
```
Comment (CIPop) [inline provisioning/service/src/ProvisioningServiceClient.cs]: We should remove this ctor - CancellationToken makes sense only for async APIs.

File: provisioning/transport/amqp/src/AmqpClientConnection.cs [hunk 3]
```diff
@@ -123,44 +123,41 @@ void OnConnectionClosed(object o, EventArgs args)
...
             websocket.Options.SetBuffer(WebSocketConstants.BufferSize, WebSocketConstants.BufferSize);
 
-            // TODO: expose the Proxy setting as public API on the transport layer
             //Check if we're configured to use a proxy server
             try
             {
-                IWebProxy webProxy = WebRequest.DefaultWebProxy;
-                Uri proxyAddress = webProxy != null ? webProxy.GetProxy(websocketUri) : null;
-                if (!websocketUri.Equals(proxyAddress))
+                if (webProxy != ProvisioningTransportHandlerAmqp.s_defaultProxy)
```
Comment (CIPop) [inline provisioning/transport/amqp/src/AmqpClientConnection.cs]: Please add logging.

File: provisioning/transport/amqp/src/AmqpClientConnection.cs [hunk 4]
```diff
@@ -123,44 +123,41 @@ void OnConnectionClosed(object o, EventArgs args)
...
                 {
                     // Configure proxy server
                     websocket.Options.Proxy = webProxy;
                 }
             }
             catch (PlatformNotSupportedException)
             {
-                // .NET Core doesn't support WebProxy configuration - ignore this setting.
+                // .NET Core 2.0 doesn't support WebProxy configuration - ignore this setting.
```
Comment (CIPop) [inline provisioning/transport/amqp/src/AmqpClientConnection.cs]: Please log a warning.

File: provisioning/transport/http/src/HttpAuthStrategyTpm.cs [hunk 5]
```diff
@@ -16,15 +18,24 @@ public HttpAuthStrategyTpm(SecurityProviderTpm security)
...
         {
             var serviceCredentials = new TpmCredentials();
             var tpmDelegatingHandler = new TpmDelegatingHandler(_security);
             var apiVersionDelegatingHandler = new ApiVersionDelegatingHandler();
 
+            HttpClientHandler httpClientHandler = new HttpClientHandler();
+
+            if (proxy != ProvisioningTransportHandlerHttp.s_defaultProxy)
+            {
+                httpClientHandler.UseProxy = (proxy != null);
```
Comment (CIPop) [inline provisioning/transport/http/src/HttpAuthStrategyTpm.cs]: Please add logging.

File: provisioning/transport/http/src/HttpAuthStrategyTpm.cs [hunk 6]
```diff
@@ -16,15 +18,24 @@ public HttpAuthStrategyTpm(SecurityProviderTpm security)
...
         }
 
-        public override DeviceProvisioningServiceRuntimeClient CreateClient(Uri uri)
+        public override DeviceProvisioningServiceRuntimeClient CreateClient(Uri uri, IWebProxy proxy)
         {
             var serviceCredentials = new TpmCredentials();
             var tpmDelegatingHandler = new TpmDelegatingHandler(_security);
             var apiVersionDelegatingHandler = new ApiVersionDelegatingHandler();
 
+            HttpClientHandler httpClientHandler = new HttpClientHandler();
```
Comment (CIPop) [inline provisioning/transport/http/src/HttpAuthStrategyTpm.cs]: This now seems common code - please extract in the common portion, leaving only TPM/X509 particular code in these classes.

File: provisioning/transport/http/src/HttpAuthStrategyX509.cs [hunk 7]
```diff
@@ -20,13 +22,22 @@ public HttpAuthStrategyX509(SecurityProviderX509 security)
             _security = security;
         }
 
-        public override DeviceProvisioningServiceRuntimeClient CreateClient(Uri uri)
+        public override DeviceProvisioningServiceRuntimeClient CreateClient(Uri uri, IWebProxy proxy)
         {
             _certificate = _security.GetAuthenticationCertificate();
 
+            HttpClientHandler httpClientHandler = new HttpClientHandler();
```
Comment (CIPop) [inline provisioning/transport/http/src/HttpAuthStrategyX509.cs]: This now seems common code - please extract in the common portion, leaving only TPM/X509 particular code in these classes.

File: provisioning/transport/mqtt/src/ProvisioningTransportHandlerMqtt.cs [hunk 8]
```diff
@@ -242,6 +245,20 @@ private static DeviceRegistrationResult ConvertToProvisioningRegistrationResult(
             websocket.Options.AddSubProtocol(WsMqttSubprotocol);
             websocket.Options.ClientCertificates.Add(clientCertificate);
 
+            //Check if we're configured to use a proxy server
+            try
+            {
+                if (Proxy != s_defaultProxy)
+                {
+                    // Configure proxy server
+                    websocket.Options.Proxy = Proxy;
```
Comment (CIPop) [inline provisioning/transport/mqtt/src/ProvisioningTransportHandlerMqtt.cs]: Please add logging.

File: provisioning/transport/mqtt/src/ProvisioningTransportHandlerMqtt.cs [hunk 9]
```diff
@@ -242,6 +245,20 @@ private static DeviceRegistrationResult ConvertToProvisioningRegistrationResult(
...
+                }
+            }
+            catch (PlatformNotSupportedException)
+            {
+                // .NET Core 2.0 doesn't support WebProxy configuration - ignore this setting.
```
Comment (CIPop) [inline provisioning/transport/mqtt/src/ProvisioningTransportHandlerMqtt.cs]: Please add logging.

File: common/src/service/HttpClientHelper.cs [hunk 10]
```diff
@@ -43,15 +43,27 @@ sealed class HttpClientHelper : IHttpClientHelper
...
+            {
+                HttpClientHandler httpClientHandler = new HttpClientHandler();
+                httpClientHandler.UseProxy = (webProxy != null);
+                httpClientHandler.Proxy = webProxy;
+                this.httpClientObj = new HttpClient(httpClientHandler);
+            }
+            else
+            {
+                this.httpClientObj = new HttpClient();
+            }
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply common/src/service/HttpClientHelper.cs]: I could not understand this: NL after closing brace (except when followed by another })

File: e2e/test/ClientProxyE2ETests.cs [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+namespace Microsoft.Azure.Devices.E2ETests
+{
+    using TransportType = Microsoft.Azure.Devices.Client.TransportType;
+    using Http1TransportSettings = Microsoft.Azure.Devices.Client.Http1TransportSettings;
+    using AmqpTransportSettings = Microsoft.Azure.Devices.Client.AmqpTransportSettings;
+
+    [TestClass]
+    [TestCategory("IoTHub-E2E")]
+    [TestCategory("ProxyE2ETests")]
+    public partial class ClientProxyE2ETests
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply e2e/test/ClientProxyE2ETests.cs]: this class is not used anymore

File: e2e/test/ClientProxyE2ETests.cs [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    {
+        private const string DevicePrefix = "E2E_DeviceProxy_";
+        private const string ModulePrefix = "E2E_ModuleProxy_";
+        private const string DevicePrefixTimeout = "E2E_DeviceProxy_Timeout_";
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply e2e/test/ClientProxyE2ETests.cs]: this class is not used anymore

File: e2e/test/ServiceProxyE2ETests.cs [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+        public async Task RegistryManager_AddDeviceWithSelfSignedCertificate_Http_WithProxy()
+        {
+            Http1TransportSettings httpTransportSettings = new Http1TransportSettings();
+            httpTransportSettings.Proxy = new WebProxy(ProxyServerAddress);
+
+            await RegistryManager_AddDeviceWithCertificateAuthorityAuthentication(httpTransportSettings).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        public async Task RegistryManager_UpdateDesiredProperties_Http_WithProxy()
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply e2e/test/ServiceProxyE2ETests.cs]: I have grouped the tests for Service Client, Registry Manager and Job Client under IoTHubServiceProxyE2ETests

File: e2e/test/netstandard20/ClientProxyE2ETests.NetCore.cs [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+                await moduleClient.SendEventAsync(testMessage).ConfigureAwait(false);
+            }
+            finally
+            {
+                await moduleClient.CloseAsync().ConfigureAwait(false);
+            }
+        }
+
+        #region Helper Functions
+        private async Task<PartitionReceiver> CreateEventHubReceiver(string deviceName)
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply e2e/test/netstandard20/ClientProxyE2ETests.NetCore.cs]: this class is also not used anymore

File: e2e/test/netstandard20/ProvisioningTests.cs [hunk 15]
```diff
@@ -47,93 +52,135 @@ public enum X509EnrollmentType
...
         {
-            await ProvisioningDeviceClient_ValidRegistrationId_Register_Ok(nameof(ProvisioningTransportHandlerMqtt), nameof(SecurityProviderX509Certificate), X509EnrollmentType.Group, TransportFallbackType.WebSocketOnly).ConfigureAwait(false);
+            await ProvisioningDeviceClient_ValidRegistrationId_Register_Ok(nameof(ProvisioningTransportHandlerMqtt), nameof(SecurityProviderX509Certificate), X509EnrollmentType.Group, TransportFallbackType.WebSocketOnly, null).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        [TestCategory("ProxyE2ETests")]
+        public async Task ProvisioningDeviceClient_ValidRegistrationId_HttpWithProxy_Tpm_RegisterOk()
+        {
+            await ProvisioningDeviceClient_ValidRegistrationId_Register_Ok(nameof(ProvisioningTransportHandlerHttp), nameof(SecurityProviderTpmHsm), null, null, ProxyServerAddress).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply e2e/test/netstandard20/ProvisioningTests.cs]: Previously HttpAuthStrategy was different for both Tpm and X509. As per feedback, I have extracted the common code to HttpAuthStrategy and removed duplicate tests

File: e2e/test/netstandard20/ProvisioningTests.cs [hunk 16]
```diff
@@ -373,6 +442,66 @@ public async Task ProvisioningDeviceClient_InvalidGlobalAddress_Register_MqttWs_
             }
         }
 
+        [TestMethod]
+        [TestCategory("ProxyE2ETests")]
+        public async Task ProvisioningServiceClient_Tpm_IndividualEnrollments_Query_HttpWithProxy_Ok()
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply e2e/test/netstandard20/ProvisioningTests.cs]: Done

File: iothub/service/src/AmqpTransportSettings.cs [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+using System.Collections.Generic;
+using System.Net;
+using System.Text;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// contains Http1 transport-specific settings for Service Client
+    /// </summary>
+    public sealed class AmqpTransportSettings
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply iothub/service/src/AmqpTransportSettings.cs]: this is now removed

File: common/src/service/HttpClientHelper.cs [hunk 18]
```diff
@@ -43,15 +43,24 @@ sealed class HttpClientHelper : IHttpClientHelper
...
-            this.httpClientObj = new HttpClient();
+            IWebProxy webProxy = transportSettings.Proxy;
+            HttpClientHandler httpClientHandler = new HttpClientHandler();
+            if (webProxy != Http1TransportSettings.s_defaultProxy)
+            {
+                httpClientHandler.UseProxy = (webProxy != null);
+                httpClientHandler.Proxy = webProxy;
+            }
+
+            this.httpClientObj = new HttpClient(httpClientHandler);
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply common/src/service/HttpClientHelper.cs]: Done

File: iothub/device/src/AmqpTransportSettings.cs [hunk 19]
```diff
@@ -15,6 +17,7 @@ public sealed class AmqpTransportSettings : ITransportSettings
         static readonly TimeSpan DefaultOperationTimeout = TimeSpan.FromMinutes(1);
         static readonly TimeSpan DefaultOpenTimeout = TimeSpan.FromMinutes(1);
         const uint DefaultPrefetchCount = 50;
+        internal static readonly IWebProxy s_defaultProxy = new DefaultWebProxySettings();
```
[the remark this answers is not in the document]
Comment (abhipsaMisra) [reply iothub/device/src/AmqpTransportSettings.cs]: If it is public property, then the API surface will change again. Is that ok?

File: common/src/DefaultWebProxySettings.cs [hunk 20]
```diff
@@ -0,0 +1,23 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Shared;
+using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Text;
+
+namespace Microsoft.Azure.Devices.Client.Common
```
Comment (CIPop) [inline common/src/DefaultWebProxySettings.cs]: I don't think we need another common namespace but if you think we do, it should be outside of Client (it's used everywhere).
Logging is using `Shared`.

File: e2e/test/MessageE2ETests.cs [hunk 21]
```diff
@@ -606,6 +666,51 @@ private async Task VerifyReceivedC2DMessage(Client.TransportType transport, Devi
...
+                    transportSettings = new ITransportSettings[] { new Client.Transport.Mqtt.MqttTransportSettings(transport) };
+                    break;
+
+                case Client.TransportType.Http1:
+                    transportSettings = new ITransportSettings[] { new Client.Http1TransportSettings() };
+                    break;
+
+                default:
+                    throw new NotSupportedException($"Unknown transport: '{transport}'.");
+            }
```
Comment (CIPop) [inline e2e/test/MessageE2ETests.cs]: Nit: NL after closing brace.

File: e2e/test/netstandard20/ProvisioningTests.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (CIPop) [reply e2e/test/netstandard20/ProvisioningTests.cs]: Makes sense then. Also: thanks for simplifying!

File: iothub/device/src/IotHubConnection.cs [hunk 22]
```diff
@@ -337,27 +339,29 @@ async Task<ClientWebSocket> CreateClientWebSocketAsync(Uri websocketUri, TimeSpa
...
+                        Logging.Info(this, $"{nameof(CreateClientWebSocketAsync)} Setting ClientWebSocket.Options.Proxy");
+                }
+            }
+            catch (PlatformNotSupportedException)
             {
-                // Configure proxy server
-                websocket.Options.Proxy = webProxy;
+                // .NET Core 2.0 doesn't support proxy. Ignore this setting.
+                if (Logging.IsEnabled)
+                    Logging.Error(this, $"{nameof(CreateClientWebSocketAsync)} PlatformNotSupportedException thrown as .NET Core 2.0 doesn't support proxy");
```
Comment (CIPop) [inline iothub/device/src/IotHubConnection.cs]: Writing it this way is dangerous: either add braces or write the statement on the same line.

**Pull request merged**